src/vision/status_quo/barbara_wishes_for_easy_runtime_switch.md

```diff
@@ -0,0 +1,59 @@
+# 😱 Status quo stories: Barbara wishes for an easy runtime switch
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from
+real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async
+Rust programmers face today.
+
+## The story
+
+[Barbara] has been working on an async codebase for the past 5 years. It is extremely mature and
+quite large (in the millions of lines of code). The product has traditionally been run on the server
+only, but now they want to move large parts to the edge. After some investigation, they realize that
+the open-source async runtime they have been using (runtime X), while great for server workloads, is
+not very performant on the low-end edge devices.
+
+Runtime X makes tradeoffs in memory for performance and while performance is still a concern on
+the low-end edge devices Barbara is targetting, memory consumption is even a hard constraint there.
+Additionally, the usage of a threadpool by runtime X does not help either as the targetted edge
+devices have either a single CPU thread or only a few.
+
+They had never planned to run on edge devices and so it was assumed that they would always stick
+to runtime X. Barbara looks into the how much work it would be to convert their codebase from
+runtime X to a custom runtime tuned to their workload. To their disapointment, they discover many
+areas, where their choice of runtime permeates the code base:
+
+* They extensively use runtime-dependent helpers (e.g., a select! macro).
+* The timers, tasks, and async locks implementation all assume runtime X.
```

Comment:
I had a similar understanding/feeling but I thought maybe I'm missing something so didn't object much to these points. @nikomatsakis @rylev ?

Reply:
Well, as a simple example, `tokio::select` is a specific `select` macro, and tokio has async locks that are closely integrated with its runtime...? I believe other runtimes offer similar facilities. I think that's what I assumed was being referred to here.

Reply:
> tokio has async locks that are closely integrated with its runtime

Yeah, I thought so too but recently @taiki-e informed me that that's not correct and you can build tokio with only the `sync` module w/o the rest.

Reply:
> recently @taiki-e informed me that that's not correct and you can build tokio with only the `sync` module w/o the rest.

Yeah, they don't depend on a specific runtime.

Reply:
(Btw, @zeenix marked this conversation as resolved, but I reverted it to unresolved because the relevant text hadn't been updated and seems not correct.)

Reply:
@taiki-e thanks. I actually did but somehow that change got lost locally. :( I'll update again..

Reply:
@taiki-e actually, I did fix it: it says:

> The timers, tasks, and async locks API is all tokio-specific,

While they're not tied to the runtime itself, they **are** tokio-specific in the sense that it's all custom API and not implementing some traits from stardardized/common traits.

I also added:

>  even if they're not tied to the tokio runtime and can be used independently of it.

Reply:
> The timers, tasks, and async locks API is all tokio-specific,

Timers and tasks actually depend on runtime, so I think this is misleading.

> While they're not tied to the runtime itself, they **are** tokio-specific in the sense that it's all custom API and not implementing some traits from stardardized/common traits.

I'm not sure what you are talking about. What are the standardized/common traits used in async lock primitives? (I don't know of any other traits that are used for that purpose other than `Future`.)

Also, what do you mean by "custom API"?

> I did fix it

I don't think `select` related thing was fixed.

Reply:
Well they are runtime dependent in a way - you cannot use the `tokio::select` macro in a project that doesn't use `tokio` without bringing `tokio` in as a dependency. So it's not quite as bad as being directly dependent on the runtime in order to function, but there is still a dependency relationship.

Reply:
@rylev It would be quite misleading to rephrase "depends on the library providing the runtime" to "depends on the runtime".

Reply:
> > The timers, tasks, and async locks API is all tokio-specific,
> 
> Timers and tasks actually depend on runtime, so I think this is misleading.

Right, need to break this line then.



> > While they're not tied to the runtime itself, they **are** tokio-specific in the sense that it's all custom API and not implementing some traits from stardardized/common traits.
> 
> I'm not sure what you are talking about. What are the standardized/common traits used in async lock primitives? (I don't know of any other traits that are used for that purpose other than `Future`.)

There are none, that's the problem, not that tokio doesn't implement them. Please keep in mind that the point is not to criticize tokio here (and hence the reason we were reluctant to even mention it here).

Reply:
I took a stab at writing something below. I do think that needing a dependency on tokio is worth nothing; I also did not know that tokio-util can be used independently, and it seems like others didn't (I didn't see a lot of docs). So i think it's valid for Barbara in the story to be confused about this, but we should totally clarify the reality of the situation as well.

Reply:
I do wonder how the nifty [preemption stuff](https://tokio.rs/blog/2020-04-preemption) works with mutexes when you're not using the tokio runtime for other things! I'd like to read more into the tokio codebase...

Reply:
@rylev you *can* bring in the macros, io traits, etc... w/o the runtime by depending on tokio without specifying the runtime flag. Tokio avoids default features specifically for this case.

Reply:
When not using Tokio types w/ the tokio runtime, there is no preemption and the types behave as one would otherwise expect.